s3fs/core.py

```diff
@@ -1814,14 +1814,58 @@ def __init__(
         self.append_block = False
 
         if "a" in mode and s3.exists(path):
-            loc = s3.info(path)["size"]
+            # See:
+            #  put: https://boto3.amazonaws.com/v1/documentation/api/latest
+            #  /reference/services/s3.html#S3.Client.put_object
+
+            # head : https://boto3.amazonaws.com/v1/documentation/api/latest
+            # /reference/services/s3.html#S3.Client.head_object
+            head = {
+                k: v
+                for k, v in self._call_s3(
+                    "head_object",
+                    self.kwargs,
+                    Bucket=bucket,
+                    Key=key,
+                    **version_id_kw(version_id),
+                    **self.req_kw,
+                ).items()
+                if (
+                    k
+                    in {
+                        "CacheControl",
+                        "ContentDisposition",
+                        "ContentEncoding",
+                        "ContentLanguage",
+                        "ContentLength",
+                        "ContentType",
+                        "Expires",
+                        "WebsiteRedirectLocation",
+                        "ServerSideEncryption",
+                        "SSECustomerAlgorithm",
+                        "SSEKMSKeyId",
+                        "BucketKeyEnabled",
+                        "StorageClass",
+                        "ObjectLockMode",
+                        "ObjectLockRetainUntilDate",
+                        "ObjectLockLegalHoldStatus",
+                        "Metadata",
+                    }
+                    and k not in self.s3_additional_kwargs
+                )
+            }
+
+            loc = head.pop("ContentLength")
```

Comment:
How about smth like this (untested) to reduce levels and make it a bit more readable:

```suggestion
            # See:
            # put: https://boto3.amazonaws.com/v1/documentation/api/latest
            # /reference/services/s3.html#S3.Client.put_object
            #
            # head: https://boto3.amazonaws.com/v1/documentation/api/latest
            # /reference/services/s3.html#S3.Client.head_object
            head = self._call_s3(
                "head_object",
                self.kwargs,
                Bucket=bucket,
                Key=key,
                **version_id_kw(version_id),
                **self.req_kw,
            )
            
            head = {
                key: value
                for key, value in head.items()
                if key in PRESERVE_KWARGS and key not in self.s3_additional_kwargs
            }

            loc = head.pop("ContentLength")
```